cities/modeling/model_components.py

```diff
@@ -0,0 +1,217 @@
+from typing import Dict, Optional
+
+import pyro
+import pyro.distributions as dist
+import torch
+
+
+def get_n(categorical: Dict[str, torch.Tensor], continuous: Dict[str, torch.Tensor]):
+    N_categorical = len(categorical)
+    N_continuous = len(continuous)
+
+    n_cat = next(iter(categorical.values())).shape[0] if N_categorical > 0 else None
+    n_con = next(iter(continuous.values())).shape[0] if N_continuous > 0 else None
+
+    if N_categorical > 0 and N_continuous > 0:
+        if n_cat != n_con:
+            raise ValueError(
+                "The number of categorical and continuous data points must be the same"
+            )
+
+    n = n_cat if n_cat is not None else n_con
+
+    if n is None:
+        raise ValueError("Both categorical and continuous dictionaries are empty.")
+
+    return N_categorical, N_continuous, n
+
+
+def categorical_contribution(
+    categorical: Dict[str, torch.Tensor],
+    child_name: str,
+    leeway: float,
+    categorical_levels: Optional[Dict[str, torch.Tensor]] = None,
+) -> torch.Tensor:
+
+    categorical_names = list(categorical.keys())
+
+    if categorical_levels is None:
```

Comment:
I'm wondering if we can make this required? Otherwise it seems really easy to have mismatched categories if you're either conditioning or forward simulating subsets of your categories.